### ISSUE TITLE
Enhancement: option to override SSH connection IP

### ISSUE THREAD

**Issue by lesinigo** [jenkinsci/docker-plugin#743]
# TL;DR

There should be a configuration option, per every "docker cloud" in Jenkins, to allow manually specifying the ip/host for SSH connections to the slave (with the port number still being detected by the plugin from the Docker API).

# rationale

When using the SSH connection method and a local unix socket to talk to Docker (eg. `unix:///var/run/docker.sock`) it seems that the plugin reads the forwarded ssh port from Docker but then tries to connect to it on localhost and/or something like `0.0.0.0`.

This will fail when the Jenkins master itself is running inside a container on the same host that is spawning slave containers, so we need to instruct the plugin to direct SSH connection to the host ip address.

# why

The "Jenkins master in docker container spawning slave containers on the same host" seems to be a common target scenario and there is a lot of guides, blogs, howtos out there that make users simply open the Docker TCP socket to the network and then configure this plugin to talk to docker over TCP (I'm guessing that in this case the plugin will use the same IP for the ssh connections). Unfortunately, following this kind of guidance is literally the same as giving out root privileges on the docker host running Jeknins.

Letting the Jenkins master container access the mounted Docker unix socket is not a perfect situation but still better than opening a TCP address and a method commonly used by lots of projects that need to interact with Docker (eg. registrator, traefik, portainer, etc...)

# why not

Directing people towards opening the docker socket (either TCP o unix) should be avoided as much as possible especially when there is the possibility to get untrusted traffic to it. Like a CI/CD tool running commands against a codebase that could have come from an external repository controlled by someone else. Sysadmins should be diligent in restricting access to the socket as much as possible or completely avoiding this situation where possible (eg. by using JNLP method instead of SSH one in this plugin)

**Comment by pjdarton** [comment]
Unless I've misunderstood, this is already catered for:
Under the docker cloud config, expand "Docker cloud details..." and then expand the "Advanced..." bit to find "Docker Hostname or IP address".

While I've not checked the code to be 100% sure (as I've just got back from a 2 week absence and I'm wading though a huge backlog of emails), I think this is what tells Jenkins what IP address it should be connecting to in order to find the docker containers open ports, i.e. the IP address the docker daemon is exposing those ports on.

Note: If this is the case then I guess your RFE is actually highlighting a deficiency in the online help - feel free to suggest improvements so we can help make it clearer for everyone...

**Comment by pjdarton** [comment]
Closing due to lack of response
